owasp-top10-2017-apps/a3/snake-pro/app/api/pages.go

```diff
@@ -1,23 +1,70 @@
 package api
 
 import (
+    "fmt"
 	"net/http"
+    "os"
 
+    jwt "github.com/dgrijalva/jwt-go"
 	"github.com/labstack/echo"
 )
 
+func CheckUserAuth(c echo.Context) (map[string]interface{}, error) {
+    cookie, err := ReadCookie(c)
+    if err != nil {
+        return nil, fmt.Errorf("malformed token: %v", err)
+    }
+
+    token, err := jwt.Parse(string(cookie), func(token *jwt.Token) (interface{}, error) {
+        // Don't forget to validate the alg is what you expect:
+        if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+            return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
+        }
+
+        // hmacSampleSecret is a []byte containing your secret, e.g. []byte("my_secret_key")
+        return []byte(os.Getenv("SECRET_KEY")), nil
+    })
+
+    if token == nil {
+        return nil, fmt.Errorf("malformed token: %v", err)
+    }
+
+    claims, ok := token.Claims.(jwt.MapClaims)
+    if !ok {
+        return nil, fmt.Errorf("Missing claims")
+    }
+
+    return claims, nil
+}
+
 // PageLogin renders login page
 func PageLogin(c echo.Context) error {
+    auth, err := CheckUserAuth(c)
+    if err != nil {
+        return c.Redirect(302, "/login")
```

Comment:
Ops, I think we have a loop here! 😅 After running your app and accessing `/login`, I got 47 redirects. What happens is that a new user do not have any authentication yet and he/she will always enter this if condition.

Reply:
Ops! That's what happen when you copy-paste code and don't have a proper test suit :/

Pushed some improvements.